src/index.ml

```diff
@@ -409,56 +512,86 @@ module Make (K : Key) (V : Value) (IO : IO) = struct
-    let log_array =
-      let compare_entry e e' = compare e.key_hash e'.key_hash in
-      let b = Array.make (Tbl.length log.mem) witness in
-      Tbl.fold
-        (fun key value i ->
-          b.(i) <- { key; value; key_hash = K.hash key };
-          i + 1)
-        log.mem 0
-      |> ignore;
-      Array.fast_sort compare_entry b;
-      b
-    in
-    let fan_size =
-      match t.index with
-      | None -> Tbl.length log.mem
-      | Some index ->
-          (Int64.to_int (IO.offset index.io) / entry_size) + Tbl.length log.mem
+    flush_instance t;
+    Mutex.lock t.merge_lock;
+    Log.info (fun l -> l "[%s] merge" (Filename.basename t.root));
+    let log_async =
+      let io =
+        let log_async_path = log_async_path t.root in
+        IO.v ~fresh:true ~readonly:false ~generation:(Int64.succ t.generation)
+          ~fan_size:0L log_async_path
+      in
+      let mem = Tbl.create 0 in
+      { io; mem }
     in
-    let fan_out = Fan.v ~hash_size:K.hash_size ~entry_size fan_size in
-    let merge =
-      IO.v ~readonly:false ~fresh:true ~generation
-        ~fan_size:(Int64.of_int (Fan.exported_size fan_out))
-        merge_path
+    t.log_async <- Some log_async;
+
+    let go () =
+      let log = assert_and_get t.log in
+      let generation = Int64.succ t.generation in
+      let log_array =
+        let compare_entry e e' = compare e.key_hash e'.key_hash in
+        let b = Array.make (Tbl.length log.mem) witness in
+        Tbl.fold
+          (fun key value i ->
+            b.(i) <- { key; key_hash = K.hash key; value };
+            i + 1)
+          log.mem 0
+        |> ignore;
+        Array.fast_sort compare_entry b;
+        b
+      in
+      let fan_size =
+        match t.index with
+        | None -> Tbl.length log.mem
+        | Some index ->
+            (Int64.to_int (IO.offset index.io) / entry_size)
+            + Array.length log_array
+      in
+      let fan_out = Fan.v ~hash_size:K.hash_size ~entry_size fan_size in
+      let merge =
+        let merge_path = merge_path t.root in
+        IO.v ~fresh:true ~readonly:false ~generation
+          ~fan_size:(Int64.of_int (Fan.exported_size fan_out))
+          merge_path
+      in
+      let index =
+        match t.index with
+        | None ->
+            let io =
+              IO.v ~fresh:true ~readonly:false ~generation ~fan_size:0L
+                (index_path t.root)
+            in
+            append_remaining_log fan_out log_array 0 merge;
+            { io; fan_out }
+        | Some index ->
+            let index = { index with fan_out } in
+            merge_with log_array index merge;
+            index
+      in
+      Fan.finalize index.fan_out;
+      IO.set_fanout merge (Fan.export index.fan_out);
+      Mutex.lock t.rename_lock;
+      IO.rename ~src:merge ~dst:index.io;
+      t.index <- Some index;
+      let log_async = assert_and_get t.log_async in
+      IO.clear ~keep_generation:true log.io;
```

Comment:
Why keep the generation in `log.io` here, since you change it two lines below? 
I think that you don't need to have the generation in `log_async`, no?

Reply:
We have to keep in mind that a readonly instance might read this at any time. If a readonly instance reads the index while the generation is set to zero (which happens if we don't keep the generation because it will be reset), it will believe that our index is at generation 0 and in particular won't read the `index` (but perhaps we can change this behaviour if that doesn't seem right).

Reply:
And about where the generation is needed, strictly speaking we are only reading the generation in the `log` file, but having this header meaningful everywhere can help debugging, because it allows us to understand at what stage of a merge (or not) we are. That can be useful when writing a diagnosis tool.
I don't think this part of the code involves the generation in `log_async` however.